spacy/tokens/doc.pyx

```diff
@@ -1462,6 +1462,146 @@ cdef class Doc:
+            raise ValueError(Errors.E1017)
+        if token_annotations[POS] is not None:
+            for pp in set(token_annotations[POS]):
+                if pp not in parts_of_speech.IDS:
+                    raise ValueError(Errors.E1021.format(pp=pp))
+
+        attrs = []
+        annotations = [token_annotations[key] for key in annot_types]
+        possible_headings = [POS, HEAD, DEP, LEMMA, TAG, MORPH]
+        for a, annot in enumerate(annotations):
+            if annot is not None:
+                if len(annot) != len(words):
+                    raise ValueError(Errors.E189)
+                attrs.append(possible_headings[a])
+
+        # If there are any other annotations, set them.
+        if attrs:
+            array = self.to_array(attrs)
+            if array.ndim == 1:
+                array = numpy.reshape(array, (array.size, 1))
+            j = 0
+
+            for annot in annotations:
+                if annot:
+                    if annot is token_annotations[HEAD]:
+                        for i in range(len(words)):
+                            array[i, j] = annot[i]
+                    elif annot is token_annotations[MORPH]:
+                        for i in range(len(words)):
+                            array[i, j] = self.vocab.morphology.add(token_annotations[MORPH][i])
+                    else:
+                        for i in range(len(words)):
+                            array[i, j] = self.vocab.strings.add(annot[i])
+                    j += 1
+            self.from_array(attrs, array)
+
+        ### Span/document properties ###
+
+        # Complement other document-level properties (cats, spans, ents).
+        self.cats = doc_json.get("cats", self.cats)
+
+        # Set sentence boundaries, if dependency parser not available but sentences are specified in JSON.
+        if not self.has_annotation("DEP"):
+            for sent in doc_json.get("sents", {}):
+                for t in self.char_span(sent["start"], sent["end"]):
+                    self[t.i].is_sent_start = t.i == 0
+
+        if "spans" in doc_json:
+            for span_group in doc_json.get("spans", {}):
+                spans = []
+                for span in doc_json["spans"][span_group]:
+                    char_span = self.char_span(span["start"], span["end"], span["label"], span["kb_id"])
+                    if char_span is None:
+                        raise ValueError(Errors.E1038.format(obj="span"))
+                    spans.append(char_span)
+                self.spans[span_group] = spans
+
+        if "ents" in doc_json:
+            ents = []
+            for ent in doc_json["ents"]:
+                char_span = self.char_span(ent["start"], ent["end"], ent["label"])
+                if char_span is None:
+                    raise ValueError(Errors.E1038.format(obj="entity"))
+                ents.append(char_span)
+            self.ents = ents
+
+        # Add custom attributes.
```

Comment:
We need to document better that this is only for `Doc` extensions. I suspect adding support for `Token` and `Span` extensions will be needed, but not in this PR.

Reply:
Do we have an issue for this already or should we create one?